### PR TITLE
feat(redirects): handle absolute from paths when path prefix is used

### LINKED ISSUE #12497: `createRedirect` from other domain conflicts with `pathPrefix`
## Description

I need to add redirections for the legacy domain. Redirections generate correctly when I build the site without `pathPrefix`. When the website comes with pathPrefix, the prefix added at `fromPath` URL even the given URL is a full URL.

I know that I can create a redirect file from another place and paste it. However, all redirection information is stored in the node as a frontmatter so it will be nice to create in `onCreateNode` for the data consistency.

### Steps to reproduce

I'm using netlify for this website, so the below example requires `gatsby-plugin-netlify`.

1. **Add path prefix.**
   ```js
   // gatsby-config.js
   module.exports = {
     pathPrefix: '/blog',
     // ...
   }
   ```
2. **Create redirect on `onCreateNode`.**
   ```js
   // gatsby-node.js for example
   exports.onCreateNode = ({ actions }) => {
     const { createRedirect } = actions
     createRedirect({
       fromPath: `https://www.haruair.com/blog/2683`,
       toPath: `/post/about-commit-messages`,
     })
   }
   ```
3. **Build gatsby with path prefix.**
   ```sh
   gatsby build --prefix-paths
   ```

### Expected result

In `./public/_redirects`,

```
## Created with gatsby-plugin-netlify

https://www.haruair.com/blog/2683  /blog/post/about-commit-messages  302
```

### Actual result

```
## Created with gatsby-plugin-netlify

/bloghttps://www.haruair.com/blog/2683  /blog/post/about-commit-messages  302
```

### Environment

```
  System:
    OS: macOS 10.14.2
    CPU: (8) x64 Intel(R) Core(TM) i7-4790K CPU @ 4.00GHz
    Shell: 5.3 - /bin/zsh
  Binaries:
    Node: 10.11.0 - ~/.nvm/versions/node/v10.11.0/bin/node
    Yarn: 1.12.3 - ~/.nvm/versions/node/v10.11.0/bin/yarn
    npm: 6.8.0 - ~/.nvm/versions/node/v10.11.0/bin/npm
  Languages:
    Python: 2.7.15 - /usr/local/bin/python
  Browsers:
    Chrome: 72.0.3626.121
    Firefox: 60.0.1
    Safari: 12.0.2
  npmPackages:
    gatsby: ^2.1.29 => 2.1.29 
    gatsby-image: ^2.0.31 => 2.0.31 
    gatsby-plugin-manifest: ^2.0.22 => 2.0.22 
    gatsby-plugin-netlify: ^2.0.12 => 2.0.12 
    gatsby-plugin-offline: ^2.0.24 => 2.0.24 
    gatsby-plugin-react-helmet: ^3.0.8 => 3.0.8 
    gatsby-plugin-sharp: ^2.0.26 => 2.0.26 
    gatsby-source-filesystem: ^2.0.23 => 2.0.23 
    gatsby-transformer-sharp: ^2.1.15 => 2.1.15 
  npmGlobalPackages:
    gatsby-cli: 2.4.12
```

### PR DESCRIPTION
**Pull request by edykim**
## Description

`toPath` in `createRedirect` handles path prefix properly, however, `fromPath` doesn't. I added the code for supporting the `fromPath` cases.

Also, I added test cases of `createRedirect` action for the future.

## Related Issues

Fixes #12497

### CODE REVIEW

File: packages/gatsby/src/redux/actions.js [hunk 1]
```diff
@@ -1134,15 +1134,22 @@ actions.createRedirect = ({
...
+  const fromPathPrefix =
+    parsedFromPath.protocol != null || isFromPathRelativeProtocol
+      ? ``
+      : pathPrefix
+
+  const parsedToPath = url.parse(toPath)
+  const isToPathRelativeProtocol = toPath.startsWith(`//`)
   const toPathPrefix =
-    parsed.protocol != null || isRelativeProtocol ? `` : pathPrefix
+    parsedToPath.protocol != null || isToPathRelativeProtocol ? `` : pathPrefix
```
Comment (pieh) [inline packages/gatsby/src/redux/actions.js]: Can we abstract path normalization to function that would be used by both `fromPath` and `toPath` to not repeat same code twice?

something like:
```js
/**
 * Check if path is absolute and add pathPrefix in front if it's not
 */
const maybeAddPathPrefix = (path, pathPrefix) => {
  const parsedToPath = url.parse(path)
  const isPathRelativeProtocol = path.startsWith(`//`)
  return `${parsedToPath.protocol != null || isPathRelativeProtocol ? `` : pathPrefix}${path}`
}
```
And then 

```js
  return {
    type: `CREATE_REDIRECT`,
    payload: {
      fromPath: maybeAddPathPrefix(fromPath, pathPrefix),
      isPermanent,
      redirectInBrowser,
      toPath: maybeAddPathPrefix(toPath, pathPrefix),
      ...rest,
    },
  }
```

File: packages/gatsby/src/redux/__tests__/redirects.js [hunk 2]
```diff
@@ -0,0 +1,139 @@
+"use strict"
```
Comment (pieh) [inline packages/gatsby/src/redux/__tests__/redirects.js]: Is this needed?

Comment (edykim) [reply packages/gatsby/src/redux/__tests__/redirects.js]: I removed it :)

File: packages/gatsby/src/redux/actions.js (hunk 1 above)
Comment (edykim) [reply packages/gatsby/src/redux/actions.js]: Thanks for the review. I updated the code.

**Comment by edykim** [comment]
Thanks for the thoughtful review. I updated the code based on your recommendation. Please review the changes when you have time. 😊

**Pull request merged**